src/metabase/util/analytics.clj

```diff
@@ -0,0 +1,119 @@
+  "An instance of a Snowplow emitter"
+  (delay
+    (let [client (-> (HttpClients/custom)
+                     (.setConnectionManager (PoolingHttpClientConnectionManager.))
+                     (.build))
+          builder (-> (ApacheHttpClientAdapter/builder)
+                      (.httpClient client)
+                      (.url (snowplow-url)))
+          adapter (.build ^ApacheHttpClientAdapter$Builder builder)
+          batch-emitter-builder (-> (BatchEmitter/builder)
+                                    (.bufferSize 1)
+                                    (.httpClientAdapter adapter))]
+      (.build ^BatchEmitter$Builder batch-emitter-builder))))
+
+(def ^:private ^Tracker tracker
+  "An instance of a Snowplow tracker"
+  (delay
+   (-> (Tracker$TrackerBuilder. ^Emitter @emitter "sp" "metabase")
+       (.build))))
+
+(defn- subject
+  "Create a Subject object for a given user ID, to be included in analytics events"
+  [user-id]
+  (-> (Subject$SubjectBuilder.)
+      (.userId (str user-id))
+      (.build)))
+
+(defn- context
+  "Common context included in every analytics event"
+  []
+  (new SelfDescribingJson
+       "iglu:com.metabase/instance/jsonschema/1-0-0"
+       {"id"             (public-settings/analytics-uuid),
+        "version"        {"tag" (:tag (public-settings/version))},
+        "token-features" (into {} (for [[token enabled?] (public-settings/token-features)]
+                                    [(name token) enabled?]))}))
+
+(defn- payload
+  "A SelfDescribingJson object containing the provided event data, which can be included as the payload for an
+  analytics event"
+  [schema version event-data]
+  (new SelfDescribingJson
+       (format "iglu:com.metabase/%s/jsonschema/%s" (name schema) version)
+       ;; Make sure keywords are converted to strings
+       (into {} (for [[k v] event-data] [(name k) (if (keyword? v) (name v) v)]))))
+
+(defn- track-schema-event
+  "Send a single analytics event to the Snowplow collector, if tracking is enabled for this MB instance"
+  [schema version user-id event-data]
+  (when (public-settings/anon-tracking-enabled)
+    (try
+      (let [^Unstructured$Builder builder (-> (. Unstructured builder)
+                                              (.eventData (payload schema version event-data))
+                                              (.customContext [(context)]))
+            ^Unstructured$Builder builder' (if user-id
+                                             (.subject builder (subject user-id))
+                                             builder)
+            ^Unstructured event (.build builder')]
+        (.track ^Tracker @tracker event))
+      (catch Throwable e
+        (log/debug e (trs "Error sending Snowplow analytics event {0}" (name (:event event-data))))))))
+
+;; Snowplow analytics interface
+
+(defmulti track-event
+  "Send a single analytics event to Snowplow"
+  (fn [event & _] event))
+
+(defmethod track-event :new_instance_created
```

Comment:
why not just use `:lisp-case` keywords for all these events and convert them to `snake_case` before sending the data off to Snowplow or whatever? Ideally I think we wouldn't have any `snake_case` in our Clojure code and just have things convert automatically where appropriate... in Toucan 2 I have some stuff set up to do that so hopefully in the near future we'll be eliminating the weirdness there for database interaction at least

Reply:
Actually namespacing these keywords might let you deduplicate stuff a bit. For example, both `:dashboard_created` and `:question_added_to_dashboard` use the `:dashboard` schema, right? If you named these keywords something like `::dashboard-created` and `::question-added-to-dashboard` and derived them both from `::dashboard` then maybe they could share an impl of `track-event`. By namespacing the keywords you get the added bonus of knowing where those keywords come from. You could still just do `name` to drop the namespace from the keyword and then `metabase.util/snake-key` or something to convert to `snake_case`

Reply:
I tried your suggestion of deriving event keywords from schema keywords, but then I realized the multimethod wasn't necessary anymore since the implementations could all be the same. So I got rid of the keyword hierarchy and just encoded the event -> schema mapping in a map directly.